pkg/proxy/apis/proxyconfig/scheme/scheme.go

```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheme
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/kubernetes/pkg/proxy/apis/proxyconfig"
+	"k8s.io/kubernetes/pkg/proxy/apis/proxyconfig/v1alpha1"
+)
+
+// Utility functions for the Kube-proxy's proxyconfig API group
+
+// NewSchemeAndCodecs is a utility funciton that returns a Scheme and CodecFactory
+// that understand the types in the proxyconfig API group.
+func NewSchemeAndCodecs() (*runtime.Scheme, *serializer.CodecFactory, error) {
```

Comment:
I think it might make more sense to have this mirror the autogenerated scheme files, such as https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/scheme/register.go, i.e.
```go
var Scheme = runtime.NewScheme()
var Codecs = serializer.NewCodecFactory(Scheme)
var ParameterCodec = runtime.NewParameterCodec(Scheme)

func init() {
	v1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
	AddToScheme(Scheme)
}

func AddToScheme(scheme *runtime.Scheme) {
  proxyconfig.AddToScheme(scheme)
  v1alpha1.AddToScheme(scheme)
}
```

WDYT @sttts?

Reply:
I agree, this looks looks better.

Reply:
I know that `foo.AddToScheme()` can return an error and we're ignoring it in the generated code. Worth fixing eventually?

Reply:
This PR follows the example of #50789 and the func was borrowed from there directly. I didn't modify it. @mtaufen WDYT?

https://github.com/kubernetes/kubernetes/pull/50789/files#diff-6399b18fd5d8881fa9cabea897953189

Reply:
I guess the discussion on #50718 and #50855 never really wrapped up. I have a general dislike for globals... If we did a global Scheme/Codecs per API group, I'd probably be ok with that because it's mostly equivalent to NewSchemeAndCodecs but more efficient because it caches the result. Having a global Scheme that combines API groups, I'm less sure about. @mikedanese?

Reply:
> Having a global Scheme that combines API groups, I'm less sure about

@mtaufen which groups do you mean?

Reply:
We either need to accept `NewSchemeAndCodecs` as-is for the moment (given that the `kubeletconfig` api is doing that), or decide to go a different way. @deads2k @liggitt @sttts @smarterclayton thoughts? xref #50855 as well.

Reply:
```
	admissionregistrationv1alpha1.AddToScheme(scheme)
	appsv1beta1.AddToScheme(scheme)
	appsv1beta2.AddToScheme(scheme)
	appsv1.AddToScheme(scheme)
	authenticationv1.AddToScheme(scheme)
	authenticationv1beta1.AddToScheme(scheme)
	authorizationv1.AddToScheme(scheme)
	authorizationv1beta1.AddToScheme(scheme)
	autoscalingv1.AddToScheme(scheme)
	autoscalingv2beta1.AddToScheme(scheme)
	batchv1.AddToScheme(scheme)
	batchv1beta1.AddToScheme(scheme)
	batchv2alpha1.AddToScheme(scheme)
	certificatesv1beta1.AddToScheme(scheme)
	corev1.AddToScheme(scheme)
	extensionsv1beta1.AddToScheme(scheme)
	networkingv1.AddToScheme(scheme)
	policyv1beta1.AddToScheme(scheme)
	rbacv1.AddToScheme(scheme)
	rbacv1beta1.AddToScheme(scheme)
	rbacv1alpha1.AddToScheme(scheme)
	schedulingv1alpha1.AddToScheme(scheme)
	settingsv1alpha1.AddToScheme(scheme)
	storagev1beta1.AddToScheme(scheme)
	storagev1.AddToScheme(scheme)
```
@ncdc From your client-go reference, each of the items in this list is an API group. Lumping them all under the same scheme is what I mean by having a scheme that combines API groups. The difference between this vs. 1:1 Scheme:Group is that in the latter you just import schemes as needed, while in the former you have to add a group to the global scheme to use it.

Reply:
Well, not quite each, some are different versions of the same group, but yeah

Reply:
@mtaufen my apologies, I was mixing examples. The `kubeproxyconfig` scheme would only have its types added to it. It would not include any other apigroups.

Reply:
i.e. what I pasted [here](https://github.com/kubernetes/kubernetes/pull/53645#discussion_r144019395) is all you would have.

Reply:
That seems reasonable to me.